driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -16,168 +16,125 @@
 package com.datastax.driver.core;
 
 import java.net.InetSocketAddress;
-import java.nio.ByteBuffer;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 
 import com.codahale.metrics.Timer;
+import com.google.common.collect.Sets;
+import io.netty.util.HashedWheelTimer;
+import io.netty.util.Timeout;
+import io.netty.util.TimerTask;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.datastax.driver.core.exceptions.*;
-import com.datastax.driver.core.policies.*;
+import com.datastax.driver.core.policies.RetryPolicy;
 import com.datastax.driver.core.policies.RetryPolicy.RetryDecision.Type;
+import com.datastax.driver.core.policies.SpeculativeExecutionPolicy.SpeculativeExecutionPlan;
 
 /**
  * Handles a request to cassandra, dealing with host failover and retries on
  * unavailable/timeout.
  */
-class RequestHandler implements Connection.ResponseCallback {
+class RequestHandler {
     private static final Logger logger = LoggerFactory.getLogger(RequestHandler.class);
 
+    final String id;
+
     private final SessionManager manager;
     private final Callback callback;
 
     private final Iterator<Host> queryPlan;
+    private final SpeculativeExecutionPlan speculativeExecutionPlan;
+    private final Set<SpeculativeExecution> runningExecutions = Sets.newHashSet();
+    private final Set<Timeout> scheduledExecutions = Sets.newHashSet();
```

Comment:
Shouldn't these sets be concurrent / externally synchronized? In particular, there are iterations over them in `cancelPendingExecutions()` which, as long as I could see, could be executed while other threads are modifying the sets.